### PR TITLE
202109 Turn off after changing inputs

### PR DESCRIPTION
**Pull request by owenThurm**
If we don't turn off the bot when a user changes inputs we enter an error loop. (For instance, if they change their username, and it's now invalid then they won't be able to continue to edit it because the pop-up will keep appearing)

We also don't want the bot to start when they're in the middle of editing their password or something, they should fully edit their settings and then turn it on again. (We did the same for the web-app version)

### CODE REVIEW

File: src/GUI/bot_gui.py [hunk 1]
```diff
@@ -154,34 +171,35 @@ def get_validated_inputs(self) -> Optional[Tuple]:
...
+        ) = validated_data
 
-        if (
-            username
-            and password
-            and num_comments
-            and self.targets_list
-            and self.validate_user(username)
-        ):
+        if username and password and num_comments and targets_list:
```
Comment (DanielGelfand) [inline src/GUI/bot_gui.py:202]: Note to self: Data already validated at this point.

File: src/constants.py [hunk 2]
```diff
@@ -44,6 +44,8 @@
 SEND_TO_PLACEHOLDER = "Your brand IG username..."
 TARGET_PLACEHOLDER = "Target brand"
 
+WRITE_LISTENER = "write"
```
Comment (DanielGelfand) [inline src/constants.py:47]: Note to self: Trace mode

File: src/GUI/bot_gui.py [hunk 3]
```diff
@@ -234,8 +252,12 @@ def schedule_next_run_time(self, force_sleep: bool = False) -> None:
 
         write_run_time_and_runs(self.time_to_run, self.runs_today)
 
+    def on_entry_change(self, *args):
```
Comment (DanielGelfand) [inline src/GUI/bot_gui.py]: Type hint needed here?

File: src/GUI/Components/EntryWithPlaceholder.py [hunk 4]
```diff
@@ -42,3 +43,7 @@ def foc_in(self, *args: Any) -> None:
     def foc_out(self, *args: Any) -> None:
         if not self.get():
             self.put_placeholder()
+
+    def uncolor(self) -> None:
```
Comment (DanielGelfand) [inline src/GUI/Components/EntryWithPlaceholder.py:47]: Useful

File: src/GUI/Components/EntryWithPlaceholder.py [hunk 5]
```diff
@@ -28,7 +28,8 @@ def __init__(
         self.bind("<FocusIn>", self.foc_in)
         self.bind("<FocusOut>", self.foc_out)
 
-        self.put_placeholder()
+        if not self.get():
```
Comment (DanielGelfand) [inline src/GUI/Components/EntryWithPlaceholder.py:31]: Isn't this the same as foc_out?

File: src/GUI/bot_gui.py [hunk 6]
```diff
@@ -105,21 +122,21 @@ def run(self) -> None:
 
     def get_inputs(self) -> Tuple:
         username = (
-            self.username_entry.get()
-            if self.username_entry.get() != USERNAME_PLACEHOLDER
+            self.username.get()
+            if self.username and self.username.get() != USERNAME_PLACEHOLDER
```
Comment (DanielGelfand) [inline src/GUI/bot_gui.py:126]: Nice makes sure there's an input in the entry box.

File: src/GUI/bot_gui.py [hunk 7]
```diff
@@ -234,8 +252,12 @@ def schedule_next_run_time(self, force_sleep: bool = False) -> None:
 
         write_run_time_and_runs(self.time_to_run, self.runs_today)
 
+    def on_entry_change(self, *args: Any) -> None:
+        if self.is_running:
+            self.toggle_run_switch(force_off=True)
```
Comment (DanielGelfand) [inline src/GUI/bot_gui.py:257]: Note to self: Turns switch off on write

File: src/GUI/bot_gui.py (hunk 1 above)
Comment (owenThurm) [reply src/GUI/bot_gui.py:202]: WIll add all these notes as comments, they should be helpful in the future

File: src/GUI/bot_gui.py (hunk 3 above)
Comment (owenThurm) [reply src/GUI/bot_gui.py]: Added

File: src/GUI/Components/EntryWithPlaceholder.py (hunk 5 above)
Comment (owenThurm) [reply src/GUI/Components/EntryWithPlaceholder.py:31]: nice catch

File: src/GUI/Components/EntryWithPlaceholder.py [hunk 8]
```diff
@@ -28,8 +28,7 @@ def __init__(
         self.bind("<FocusIn>", self.foc_in)
         self.bind("<FocusOut>", self.foc_out)
 
-        if not self.get():
-            self.put_placeholder()
+        self.foc_out()
```
Comment (DanielGelfand) [inline src/GUI/Components/EntryWithPlaceholder.py:31]: Nice

**Review by DanielGelfand — APPROVED** [review]

**Pull request merged**